okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java

```diff
@@ -143,130 +299,339 @@ private static String checkOneDependency(ModuleDescriptor md, InterfaceDescripto
+                  modsEnabled.remove(from);
+                  modsEnabled.put(mdFound.getId(), mdFound);
+                  addTenantModule(tml, mdFound.getId(), from,
+                      TenantModuleDescriptor.Action.enable);
+                  return true;
+                }
+              }
+            }
+            errors.add(messages.getMessage("10201", req.moduleDescriptor.getId(), entry.getKey(),
+                req.interfaceDescriptor.getVersion(),
+                prov.interfaceDescriptor.getVersion() + "/" + prov.moduleDescriptor.getId()));
+          }
+        }
       }
     }
-    logger.info("getNextTM done null");
-    return null;
+    return false;
   }
 
+
   /**
    * Install modules with dependency checking only.
+   *
    * @param modsAvailable available modules
-   * @param modsEnabled enabled modules (for some tenant)
-   * @param tml install list with actions
-   * @param reinstall whether to re-install
-   * @return future
+   * @param modsEnabled   enabled modules (for some tenant)
+   * @param tml           install list with actions
+   * @param reinstall     whether to re-install
+   * @param maxIterations how many iterations to allow fixup of list.
    */
-  public static Future<Void> installSimulate(Map<String, ModuleDescriptor> modsAvailable,
-      Map<String, ModuleDescriptor> modsEnabled,
-      List<TenantModuleDescriptor> tml,
-      boolean reinstall) {
+  static void installMaxIterations(
+      Map<String, ModuleDescriptor> modsAvailable,
+      Map<String, ModuleDescriptor> modsEnabled, List<TenantModuleDescriptor> tml,
+      boolean reinstall, int maxIterations) {
+
     List<String> errors = new LinkedList<>();
+    Set<String> stickyModules = new HashSet<>();
     for (TenantModuleDescriptor tm : tml) {
       String id = tm.getId();
       ModuleId moduleId = new ModuleId(id);
       if (tm.getAction() == TenantModuleDescriptor.Action.enable) {
         if (!moduleId.hasSemVer()) {
           id = moduleId.getLatest(modsAvailable.keySet());
           tm.setId(id);
+          stickyModules.add(id);
         } else {
-          tm.setFixed(Boolean.TRUE);
+          stickyModules.add(id);
         }
         if (!modsAvailable.containsKey(id)) {
           errors.add(messages.getMessage("10801", id));
-        }
-        if (modsEnabled.containsKey(id) && !reinstall) {
-          tm.setAction(TenantModuleDescriptor.Action.uptodate);
+        } else {
+          if (modsEnabled.containsKey(id) && !reinstall) {
+            tm.setAction(TenantModuleDescriptor.Action.uptodate);
+          } else {
+            String product = new ModuleId(id).getProduct();
+            for (String enabledId : modsEnabled.keySet()) {
+              String enabledProduct = new ModuleId(enabledId).getProduct();
+              if (enabledProduct.equals(product)) {
+                if (stickyModules.contains(enabledId) && !id.equals(enabledId)) {
```

Comment:
Can you reduce the complexity, here we have this nesting: for - if - if-else - if-else - for - if - if. I can hardly follow what happens in these 7 level of nesting.

Reply:
yep